### PR TITLE
Fix tiny_map from_iter where one operation was being dropped

### PR DESCRIPTION
**Pull request by Thor-Bjorgvinsson**
## Motivation and Context
15th operation was being dropped when iterator was being read into a TinyMap

## Description
The 15th operation was being dropped when the if statement was caught because the operation had been popped of the iterator but hadn't been added to the vec of operations before the two iterators were being chained and collected into a 

## Testing
Added unit tests that reproduced the issue and verified that the issue is fixed

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: CHANGELOG.next.toml [hunk 1]
```diff
@@ -11,6 +11,12 @@
 # meta = { "breaking" = false, "tada" = false, "bug" = false, "target" = "client | server | all"}
 # author = "rcoh"
 
+[[smithy-rs]]
+message = "Fix tiny_map from_iter where one operation was being dropped"
```
Comment (david-perez) [inline CHANGELOG.next.toml:15]:
```suggestion
message = "Fix bug in AWS JSON 1.x routers where, if a service had more than 14 operations, the router was created without the route for the 15th operation."
```

`TinyMap` is not public so users won't be able to know what the message refers to. This one is a bit more descriptive.

**Review by david-perez — APPROVED** [review]
Thanks!

**Review by hlbarber — APPROVED** [review]

**Review by crisidev — APPROVED** [review]

**Comment by 82marbag** [comment]
@Thor-Bjorgvinsson you have probably ticked the wrong box in the checklist

**Comment by david-perez** [comment]
@Velfi Seems like [these tests](https://github.com/awslabs/smithy-rs/actions/runs/5114246199/jobs/9205507593?pr=2733#step:3:2389) are flaky. Although it's the first time I've seen them fail.

```text
running 4 tests
test test_smol_runtime_timeouts ... FAILED
test test_async_std_runtime_timeouts ... FAILED
test test_smol_runtime_retry ... ok
test test_async_std_runtime_retry ... ok

failures:

---- test_smol_runtime_timeouts stdout ----
thread 'test_smol_runtime_timeouts' panicked at 'actual = 769.061231ms, expected = 350ms', sdk/s3/tests/alternative-async-runtime.rs:128:5

---- test_async_std_runtime_timeouts stdout ----
thread 'test_async_std_runtime_timeouts' panicked at 'actual = 778.09071ms, expected = 350ms', sdk/s3/tests/alternative-async-runtime.rs:128:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    test_async_std_runtime_timeouts
    test_smol_runtime_timeouts

test result: FAILED. 2 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.60s
```

**Pull request merged**